tests/clients/test_experiment.py

```diff
@@ -348,334 +357,315 @@ def test_experiment_run_data_schema_multiple():
-        (SingleFilterOperator.GREATER_THAN, "gt"),
-        (SingleFilterOperator.GREATER_THAN_OR_EQUAL_TO, "ge"),
-    ],
+    "operator, value, expected_operator, expected_value",
+    discrete_test_cases("param-value", "param-value")
+    + continuous_test_cases(123.2, 123.2),
 )
-def test_single_filter_schema_relational_operators(
-    by, key, value, by_body, value_body, operator, operator_body
+def test_filter_schema_param(
+    operator, value, expected_operator, expected_value
 ):
-    filter = SingleFilter(by, key, operator, value)
-    expected_json = {
-        "by": by_body,
-        "key": key,
-        "operator": operator_body,
-        "value": value_body,
+    filter = ParamFilter("param-key", operator, value)
+    data = FilterSchema().dump(filter)
+    assert data == {
+        "by": "param",
+        "key": "param-key",
+        "operator": expected_operator,
+        "value": expected_value,
     }
-    data = SingleFilterSchema().dump(filter)
-    assert data == expected_json
 
 
 @pytest.mark.parametrize(
-    "by, key, by_body",
-    [
-        (SingleFilterBy.PROJECT_ID, None, "projectId"),
-        (SingleFilterBy.EXPERIMENT_ID, None, "experimentId"),
-        (SingleFilterBy.RUN_ID, None, "runId"),
-        (SingleFilterBy.DELETED_AT, None, "deletedAt"),
-        (SingleFilterBy.TAG, "tag-key", "tag"),
-        (SingleFilterBy.PARAM, "param-key", "param"),
-        (SingleFilterBy.METRIC, "metric-key", "metric"),
-    ],
+    "operator, value, expected_operator, expected_value",
+    continuous_test_cases(45.6, 45.6),
 )
-def test_single_filter_schema_defined_operator(by, key, by_body):
-    filter = SingleFilter(by, key, SingleFilterOperator.DEFINED, True)
-    expected_json = {
-        "by": by_body,
-        "key": key,
-        "operator": "defined",
-        "value": True,
+def test_filter_schema_metric(
+    operator, value, expected_operator, expected_value
+):
+    filter = MetricFilter("metric-key", operator, value)
+    data = FilterSchema().dump(filter)
+    assert data == {
+        "by": "metric",
+        "key": "metric-key",
+        "operator": expected_operator,
+        "value": expected_value,
     }
-    data = SingleFilterSchema().dump(filter)
-    assert data == expected_json
 
 
 @pytest.mark.parametrize(
-    "by, value, message",
+    "filter_type",
+    [ProjectIdFilter, ExperimentIdFilter, RunIdFilter, DeletedAtFilter],
+)
+def test_filter_schema_invalid_value_no_key(filter_type):
+    filter = filter_type(ComparisonOperator.EQUAL_TO, "invalid")
+    with pytest.raises(ValidationError):
+        FilterSchema().dump(filter)
+
+
+@pytest.mark.parametrize(
+    "filter_type, value", [(ParamFilter, None), (MetricFilter, "invalid")]
+)
+def test_filter_schema_invalid_value_with_key(filter_type, value):
+    filter = filter_type("key", ComparisonOperator.EQUAL_TO, value)
+    with pytest.raises(ValidationError):
+        FilterSchema().dump(filter)
+
```

Comment:
Should we also test the situation when key is absent for tag, param and metric filters?

Reply:
I don't think this is necessary as the namedtuple constructor will require users to pass a key for these filter types.